File: backend-spring-boot/src/main/java/no/vicx/backend/user/validation/CurrentPasswordValidator.java
package no.vicx.backend.user.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import no.vicx.backend.user.service.UserService;
import org.springframework.security.core.context.SecurityContextHolder;

public class CurrentPasswordValidator implements ConstraintValidator<CurrentPassword, String> {

    private final UserService userService;

    public CurrentPasswordValidator(UserService userService) {
        this.userService = userService;
    }

    Integer passwordMinLength;
    Integer passwordMaxLength;

    @Override
    public void initialize(CurrentPassword constraintAnnotation) {
        passwordMinLength = constraintAnnotation.minLength();
        passwordMaxLength = constraintAnnotation.maxLength();
    }

    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {
        return value == null
                || value.length() < passwordMinLength
                || value.length() > passwordMaxLength
                || userService.isValidPassword(
                SecurityContextHolder.getContext().getAuthentication().getName(),
                value);
    }
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/user/validation/CurrentPasswordValidator.kt
package no.vicx.backend.user.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import no.vicx.backend.user.service.UserService
import org.springframework.security.core.context.SecurityContextHolder

class CurrentPasswordValidator(
    private val userService: UserService,
) : ConstraintValidator<CurrentPassword, String> {
    private var passwordMinLength: Int = 0
    private var passwordMaxLength: Int = 0

    override fun initialize(constraintAnnotation: CurrentPassword) {
        passwordMinLength = constraintAnnotation.minLength
        passwordMaxLength = constraintAnnotation.maxLength
    }

    override fun isValid(
        value: String?,
        context: ConstraintValidatorContext,
    ): Boolean {
        val authentication = SecurityContextHolder.getContext().authentication

        // if any validation errors that should be handled by other validators, return true
        return if (value == null ||
            value.length < passwordMinLength ||
            value.length > passwordMaxLength ||
            authentication == null
        ) {
            true
        } else {
            userService.isValidPassword(authentication.name, value)
        }
    }
}
